### PR TITLE
Derive Proxy Config From Env Vars

### PR DESCRIPTION
**Pull request by jonathanmeier5**
#2837

*Description of changes:*

This PR fixes a bug in our proxy configuration logic. Prior to this PR, we only derived proxy configs from cluster config files.

This meant commands such as `eksctl anywhere download images` could not be used with a proxy.

This PR allows the CLI to derive a proxy config from the following env vars:
```
HTTPS_PROXY
HTTP_PROXY
NO_PROXY
```

*Testing (if applicable):*

I tested the `eksctl anywhere download images` command manually on an ec2 instance that could only communicate to the internet via a second ec2 instance that ran a proxy service.

I'm more than happy to add unit tests, I just wanted a review on the implementation first in case I was very off base.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/dependencies/factory.go [hunk 1]
```diff
@@ -167,6 +168,20 @@ func (f *Factory) WithProxyConfiguration(proxyConfig map[string]string) *Factory
 	return f
 }
 
+func (f *Factory) WithProxyConfigurationFromEnv() *Factory {
+	
+	if f.proxyConfiguration == nil {
+		proxyConfig := map[string]string{
+			"HTTP_PROXY":  os.Getenv("HTTP_PROXY"),
```
Comment (jonathanmeier5) [inline pkg/dependencies/factory.go:175]: Should these all be constants?

File: pkg/dependencies/factory.go [hunk 2]
```diff
@@ -167,6 +168,20 @@ func (f *Factory) WithProxyConfiguration(proxyConfig map[string]string) *Factory
...
+			"HTTPS_PROXY": os.Getenv("HTTPS_PROXY"),
+			"NO_PROXY":    os.Getenv("NO_PROXY"),
```
Comment (jonathanmeier5) [inline pkg/dependencies/factory.go:177]: Should I add documentation for these env vars?

File: pkg/dependencies/factory.go [hunk 3]
```diff
@@ -167,6 +168,19 @@ func (f *Factory) WithProxyConfiguration(proxyConfig map[string]string) *Factory
 	return f
 }
 
+func (f *Factory) WithProxyConfigurationFromEnv() *Factory {
+	if f.proxyConfiguration == nil {
```
Comment (g-gaston) [inline pkg/dependencies/factory.go:172]: I think we could treat this as another dependency
for example: rename `WithProxyConfiguration` to `UseProxyConfiguration`
And have this new method named `WithProxyConfiguration`

Then any dependency that wants the proxy config, will call `WithProxyConfiguration` as part of their dependencies. And at the build time, if the config hasn't been set manually already (with `UseProxyConfiguration`) then we read it from the environment

File: pkg/dependencies/factory.go (hunk 1 above)
Comment (g-gaston) [reply pkg/dependencies/factory.go:174]: Looking at this, I wonder if we should make this map part of the config in the `config` package

Comment (vivek-koppuru) [reply pkg/dependencies/factory.go:174]: Yea maybe this isn't something we need to keep in factory and have to specify for all dependencies where we want to share this logic

File: pkg/dependencies/factory.go [hunk 4]
```diff
@@ -601,7 +615,7 @@ func (f *Factory) WithHelmSecure() *Factory {
 }
 
 func (f *Factory) WithHelmInsecure() *Factory {
-	f.WithExecutableBuilder()
+	f.WithExecutableBuilder().WithProxyConfigurationFromEnv()
```
Comment (vivek-koppuru) [inline pkg/dependencies/factory.go:618]: Also, if we call this here, does it mean that the values from env can override the values specified in cluster spec, or vice versa? If it's the other way, maybe it is okay? But ideally we can read the values from the clusterspec if we anyways need to take it in as input, but this would make it more consistent with how the docker pull command would read the env vars

File: pkg/dependencies/factory.go (hunk 3 above)
Comment (jonathanmeier5) [reply pkg/dependencies/factory.go:171]: @g-gaston I *think* I did what you asked, let me know if it wasn't what you were imagining.

@vivek-koppuru re [this comment](https://github.com/aws/eks-anywhere/pull/2849#discussion_r937035140), I was concerned about the same thing.

I think that with the approach Guillermo outlines above the proxy config from the cluster config file will always take precedence over the env vars as long as `UseProxyConfiguration` is explicitly called as a dependency. I think this is what we want, right?

File: pkg/config/proxy.go [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+	"os"
+)
+
+const (
+	httpsProxyKey = "HTTPS_PROXY"
+	httpProxyKey  = "HTTP_PROXY"
+	noProxyKey    = "NO_PROXY"
+)
+
+func GetProxyConfigFromEnv() (map[string]string, error) {
```
Comment (g-gaston) [inline pkg/config/proxy.go:13]: any reason why this needs to return an error?

File: pkg/dependencies/factory.go [hunk 6]
```diff
@@ -162,11 +162,25 @@ func (f *Factory) WithRegistryMirror(mirror string) *Factory {
 	return f
 }
 
-func (f *Factory) WithProxyConfiguration(proxyConfig map[string]string) *Factory {
+func (f *Factory) UseProxyConfiguration(proxyConfig map[string]string) *Factory {
 	f.proxyConfiguration = proxyConfig
 	return f
 }
 
+func (f *Factory) WithProxyConfiguration() *Factory {
```
Comment (g-gaston) [inline pkg/dependencies/factory.go:170]: It should be already tested by I would add this explicitly to a test and assert not-nil

File: pkg/config/proxy.go (hunk 5 above)
Comment (g-gaston) [inline pkg/config/proxy.go:13]: Maybe add a unit test for this guy?

**Review by g-gaston — COMMENTED** [review]
Super clean!

File: pkg/config/proxy.go (hunk 5 above)
Comment (jonathanmeier5) [reply pkg/config/proxy.go]: Nope, removed it 👍

File: pkg/dependencies/factory.go [hunk 7]
```diff
@@ -162,11 +162,32 @@ func (f *Factory) WithRegistryMirror(mirror string) *Factory {
 	return f
 }
 
-func (f *Factory) WithProxyConfiguration(proxyConfig map[string]string) *Factory {
+func (f *Factory) UseProxyConfiguration(proxyConfig map[string]string) *Factory {
 	f.proxyConfiguration = proxyConfig
 	return f
 }
 
+func (f *Factory) GetProxyConfiguration() *map[string]string {
```
Comment (g-gaston) [inline pkg/dependencies/factory.go:170]: curious: what is this for and why would it return a pointer?

File: pkg/config/proxy.go [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+)
+
+func GetProxyConfigFromEnv() map[string]string {
+	config := map[string]string{
+		HttpsProxyKey: os.Getenv(HttpsProxyKey),
+		HttpProxyKey:  os.Getenv(HttpProxyKey),
+		NoProxyKey:    os.Getenv(NoProxyKey),
+	}
+
+	return config
```
Comment (g-gaston) [inline pkg/config/proxy.go:20]: nit:
```suggestion
	return map[string]string{
		HttpsProxyKey: os.Getenv(HttpsProxyKey),
		HttpProxyKey:  os.Getenv(HttpProxyKey),
		NoProxyKey:    os.Getenv(NoProxyKey),
	}
```

File: pkg/dependencies/factory.go (hunk 7 above)
Comment (jonathanmeier5) [reply pkg/dependencies/factory.go:170]: The `Factory` struct uses an unpublished `proxyConfiguration` attribute to track proxy information in the `dependencies` package. To test it in `dependencies_testing` I needed a way to access the attribute.

I don't mind passing by value here. Given the method's intent is introspective I think a pointer is the first thing that came to mind, but I don't need it. After a quick google it looks like returning smaller objects by value is more the norm in golang.

If you're getting at "why are we testing an unpublished attribute when go doesn't want us to", there are a few other solutions, but they all felt equally weird to me and involved more work so I went with this one. Happy to discuss though.
